### ISSUE TITLE
Support inversion of circuits

### ISSUE THREAD

**Issue by royess** [tencent-quantum-lab/tensorcircuit#26]
# Support inversion of circuits

Author
<list of authors’ GitHub id>

Status
<Draft | Accepted | Deferred | Rejected | Superseded>

Created
2022-07-03

## Abstract

Add a function to `Circuit` to support the inversion of a circuit.

## Motivation and Scope

Inversion of circuits is a basic feature for constructing quantum circuits. And it is useful in tasks like random benchmarking,

And as far as I know, the feature is supported by qiskit.

## Usage and Impact

For a Circuit object `c`, `c.inverse()` will return the inversion of it by inverting each of its gates and putting them in a reversed order.

If `c` contains parametric gates, we may offer a parameter for users to decide whether gate parameters are independent or shared.

## Related Work

Qiskit:

```
Help on method inverse in module qiskit.circuit.quantumcircuit:

inverse() -> 'QuantumCircuit' method of qiskit.circuit.quantumcircuit.QuantumCircuit instance
    Invert (take adjoint of) this circuit.
    
    This is done by recursively inverting all gates.
    
    Returns:
        QuantumCircuit: the inverted circuit
    
    Raises:
        CircuitError: if the circuit cannot be inverted.
    
    Examples:
    
        input:
    
        .. parsed-literal::
    
                 ┌───┐
            q_0: ┤ H ├─────■──────
                 └───┘┌────┴─────┐
            q_1: ─────┤ RX(1.57) ├
                      └──────────┘
    
        output:
    
        .. parsed-literal::
    
                              ┌───┐
            q_0: ──────■──────┤ H ├
                 ┌─────┴─────┐└───┘
            q_1: ┤ RX(-1.57) ├─────
                 └───────────┘
```

**Comment by refraction-ray** [comment]
This seems doable via inverting ``qir`` at least for noiseless circuit, should be done in this week.

**Comment by refraction-ray** [comment]
have implemented in v0.2.2, pls upgrade tensorcircuit to see whether ``c.inverse()`` works for your case
